File: src/test_wavefunction.c
#include <complex.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "wavefunction.h"

void test_initialization_scaling_freeing(void);
void test_realloc(void);
void test_inner_product(void);
void test_appending_slater_determinants(void);
void test_wavefunction_pauli_sum_multiplication(void);
void test_wavefunction_pauli_string_evolution(void);
void test_wavefunction_pauli_sum_evolution(void);
void test_wavefunction_speed(void);

void test_initialization_scaling_freeing(void) {
  WavefunctionC *wfn;
  WavefunctionC *normalized_wfn;
  WavefunctionC *adjoint_wfn;

  char *wfn_str;
  char *nwfn_str;
  char *awfn_str;

  double norm;

  unsigned int orbitals0[] = {0, 1, 0, 1};
  unsigned int orbitals1[] = {1, 0, 1, 0};

  SlaterDeterminantC *sdet0;
  SlaterDeterminantC *sdet1;

  sdet0 = slater_determinant_init_c(4, 1, orbitals0);
  sdet1 = slater_determinant_init_c(4, (double complex)I, orbitals1);

  wfn = wavefunction_init_c(4);
  printf("Wavefunction initialized at %p with sdets: %u, %u\n", (void *)wfn,
         sdet0->encoding, sdet1->encoding);

  wavefunction_append_slater_determinant_c(wfn, sdet0);
  wavefunction_append_slater_determinant_c(wfn, sdet1);

  wfn_str = wavefunction_to_string_c(wfn, 'k');
  printf("%s\n", wfn_str);
  free(wfn_str);

  norm = wavefunction_norm_c(wfn);
  printf("Norm: %f\n", norm);

  normalized_wfn = wavefunction_scalar_multiplication_c(wfn, 1 / norm);
  nwfn_str = wavefunction_to_string_c(normalized_wfn, 'k');
  printf("Normalized:\n");
  printf("%s\n", nwfn_str);
  free(nwfn_str);

  adjoint_wfn = wavefunction_adjoint_c(normalized_wfn);
  awfn_str = wavefunction_to_string_c(adjoint_wfn, 'b');
  printf("Adjoint:\n");
  printf("%s\n", awfn_str);
  free(awfn_str);

  free_wavefunction_c(wfn);
  free_wavefunction_c(normalized_wfn);
  free_wavefunction_c(adjoint_wfn);
}

void test_inner_product(void) {
  unsigned int orbitals0[] = {0, 1, 0, 1};
  unsigned int orbitals1[] = {1, 0, 1, 0};

  SlaterDeterminantC *sdet0;
  SlaterDeterminantC *sdet1;
  WavefunctionC *ket;
  WavefunctionC *bra;

  double complex inner_prod;

  char *bra_str;
  char *ket_str;

  sdet0 = slater_determinant_init_c(4, 1, orbitals0);
  sdet1 = slater_determinant_init_c(4, (double complex)I, orbitals1);

  ket = wavefunction_init_c(4);
  wavefunction_append_slater_determinant_c(ket, sdet0);
  wavefunction_append_slater_determinant_c(ket, sdet1);

  bra = wavefunction_adjoint_c(ket);
  inner_prod = wavefunction_multiplication_c(bra, ket);

  bra_str = wavefunction_to_string_c(bra, 'b');
  ket_str = wavefunction_to_string_c(ket, 'k');

  printf("(%s) * (%s) = %lf + %lfi\n", bra_str, ket_str, creal(inner_prod),
         cimag(inner_prod));

  free(bra_str);
  free(ket_str);
  free_wavefunction_c(bra);
  free_wavefunction_c(ket);
}

void test_appending_slater_determinants(void) {
  WavefunctionC *wfn1;
  WavefunctionC *wfn2;
  WavefunctionC *wfn3;

  unsigned int i, j, k, l;
  unsigned int orbitals[4];

  SlaterDeterminantC *sdet;

  char *wfn1_str;
  char *wfn2_str;
  char *wfn3_str;

  double complex product;

  printf("Order of Adding:\n");

  wfn1 = wavefunction_init_c(4);

  for (i = 0; i < 2; i++) {
    for (j = 0; j < 2; j++) {
      for (k = 0; k < 2; k++) {
        for (l = 0; l < 2; l++) {
          orbitals[0] = i;
          orbitals[1] = j;
          orbitals[2] = k;
          orbitals[3] = l;

          sdet = slater_determinant_init_c(4, 1, orbitals);
          wavefunction_append_slater_determinant_c(wfn1, sdet);
        }
      }
    }
  }

  wfn2 = wavefunction_init_c(4);

  for (i = 0; i < 2; i++) {
    for (j = 0; j < 2; j++) {
      for (k = 0; k < 2; k++) {
        for (l = 0; l < 2; l++) {
          orbitals[0] = l;
          orbitals[1] = k;
          orbitals[2] = j;
          orbitals[3] = i;

          sdet = slater_determinant_init_c(4, 1, orbitals);
          wavefunction_append_slater_determinant_c(wfn2, sdet);
        }
      }
    }
  }

  wfn1_str = wavefunction_to_string_c(wfn1, 'b');
  printf("%s\n\n", wfn1_str);
  free(wfn1_str);

  wfn2_str = wavefunction_to_string_c(wfn2, 'k');
  printf("%s\n\n", wfn2_str);
  free(wfn2_str);

  product = wavefunction_multiplication_c(wfn1, wfn2);
  printf("Product of %u orbitals with %u orbitals : %lf + %lfi\n\n", wfn1->s,
         wfn2->s, creal(product), cimag(product));

  free_wavefunction_c(wfn1);
  free_wavefunction_c(wfn2);

  wfn3 = wavefunction_init_c(4);
  printf("Repeated addition of the same orbitals:\n");
  printf("s before addition: %u\n", wfn3->s);

  for (i = 0; i < 4; i++) {
    unsigned int small_orbitals[] = {0, 1};
    sdet = slater_determinant_init_c(2, 1, small_orbitals);
    wavefunction_append_slater_determinant_c(wfn3, sdet);
  }

  printf("s after addition: %u\n", wfn3->s);

  wfn3_str = wavefunction_to_string_c(wfn3, 'k');
  printf("%s\n", wfn3_str);
  free(wfn3_str);

  free_wavefunction_c(wfn3);
}

void test_wavefunction_pauli_sum_multiplication(void) {
  unsigned int orbitals0[] = {0, 1, 0, 1};
  unsigned int orbitals1[] = {1, 0, 1, 0};
  unsigned int paulis0[] = {0, 2, 0, 1};
  unsigned int paulis1[] = {1, 0, 1, 0};

  SlaterDeterminantC *sdet0;
  SlaterDeterminantC *sdet1;
  WavefunctionC *wfn;
  WavefunctionC *new_wfn;

  PauliStringC *pString0;
  PauliStringC *pString1;
  PauliSumC *pSum;

  char *wfn_str;
  char *pSum_str;
  char *new_wfn_str;

  sdet0 = slater_determinant_init_c(4, 1, orbitals0);
  sdet1 = slater_determinant_init_c(4, 1, orbitals1);
  wfn = wavefunction_init_c(4);

  wavefunction_append_slater_determinant_c(wfn, sdet0);
  wavefunction_append_slater_determinant_c(wfn, sdet1);

  pString0 = pauli_string_init_as_ints_c(4, 1, paulis0);
  pString1 = pauli_string_init_as_ints_c(4, 1, paulis1);

  pSum = pauli_sum_init_c(4);
  pauli_sum_append_pauli_string_c(pSum, pString0);
  pauli_sum_append_pauli_string_c(pSum, pString1);

  new_wfn = wavefunction_pauli_sum_multiplication_c(pSum, wfn);

  wfn_str = wavefunction_to_string_c(wfn, 'k');
  pSum_str = pauli_sum_to_string_c(pSum);
  new_wfn_str = wavefunction_to_string_c(new_wfn, 'k');

  printf("(%s) * (%s) = %s\n", pSum_str, wfn_str, new_wfn_str);

  free(wfn_str);
  free(pSum_str);
  free(new_wfn_str);

  free_wavefunction_c(wfn);
  free_wavefunction_c(new_wfn);
  free_pauli_sum_c(pSum);
}

void test_wavefunction_pauli_string_evolution(void) {
  unsigned int orbitals0[] = {0, 1, 0, 1};
  unsigned int paulis0[] = {1, 1, 1, 1};

  SlaterDeterminantC *sdet0;
  WavefunctionC *wfn;
  WavefunctionC *new_wfn;

  PauliStringC *pString;

  char *wfn_str;
  char *pString_str;
  char *new_wfn_str;

  double norm;
  unsigned int t;

  sdet0 = slater_determinant_init_c(4, 1, orbitals0);
  wfn = wavefunction_init_c(4);
  wavefunction_append_slater_determinant_c(wfn, sdet0);

  pString = pauli_string_init_as_ints_c(4, (double complex)I, paulis0);

  wfn_str = wavefunction_to_string_c(wfn, 'k');

  for (t = 0; t < 100000; t++) {
    new_wfn = wavefunction_pauli_string_evolution_c(pString, wfn, 0.01);
    free_wavefunction_c(wfn);
    wfn = new_wfn;
  }

  pString_str = pauli_string_to_string_c(pString);
  new_wfn_str = wavefunction_to_string_c(wfn, 'k');
  norm = wavefunction_norm_c(wfn);

  printf("exp(0.01 * (%s)) * (%s) = %s\n", pString_str, wfn_str, new_wfn_str);
  printf("Norm: %lf\n", norm);

  free(wfn_str);
  free(pString_str);
  free(new_wfn_str);

  free_wavefunction_c(wfn);
  free_pauli_string_c(pString);
}

void test_wavefunction_pauli_sum_evolution(void) {
  unsigned int orbitals0[] = {0, 1, 0, 1};
  unsigned int paulis0[] = {1, 1, 1, 1};

  SlaterDeterminantC *sdet0;
  WavefunctionC *wfn;
  WavefunctionC *new_wfn;

  PauliStringC *pString0;
  PauliStringC *pString1;
  PauliSumC *pSum;

  char *wfn_str;
  char *pSum_str;
  char *new_wfn_str;

  double norm;
  unsigned int t;

  sdet0 = slater_determinant_init_c(4, 1, orbitals0);
  wfn = wavefunction_init_c(4);
  wavefunction_append_slater_determinant_c(wfn, sdet0);

  pString0 = pauli_string_init_as_ints_c(4, (0.5 * (double complex)I), paulis0);
  pString1 = pauli_string_init_as_ints_c(4, (0.5 * (double complex)I), paulis0);

  pSum = pauli_sum_init_c(4);
  pauli_sum_append_pauli_string_c(pSum, pString0);
  pauli_sum_append_pauli_string_c(pSum, pString1);

  wfn_str = wavefunction_to_string_c(wfn, 'k');

  for (t = 0; t < 100000; t++) {
    new_wfn = wavefunction_pauli_sum_evolution_c(pSum, wfn, 0.01);
    free_wavefunction_c(wfn);
    wfn = new_wfn;
  }

  pSum_str = pauli_sum_to_string_c(pSum);
  new_wfn_str = wavefunction_to_string_c(wfn, 'k');
  norm = wavefunction_norm_c(wfn);

  printf("exp(0.01 * (%s)) * (%s) = %s\n", pSum_str, wfn_str, new_wfn_str);
  printf("Norm: %lf\n", norm);

  free(wfn_str);
  free(pSum_str);
  free(new_wfn_str);

  free_wavefunction_c(wfn);
  free_pauli_sum_c(pSum);
}

void test_wavefunction_speed(void) {
  const unsigned int N = 20;
  unsigned int num_sdets = 2;
  const unsigned int num_operations = 20;

  unsigned int i, j, k;
  unsigned int orbitals[N];
  unsigned int paulis[N];
  unsigned int pauli_type;

  SlaterDeterminantC *sdet;
  WavefunctionC *wfn;
  WavefunctionC *new_wfn;

  PauliSumC *pauli_sums[num_operations];
  PauliStringC *pString;

  WavefunctionC **old_wfns;
  char *pSum_str;

  double complex coef;
  double norm;
  clock_t start_time, end_time;
  double elapsed_time;

  printf("\nRunning Wavefunction Speed Test with %u qubits, %u Slater "
         "determinants, %u Pauli sum operations.\n",
         N, num_sdets, num_operations);

  wfn = wavefunction_init_c(N);

  for (i = 0; i < num_sdets; i++) {
    for (j = 0; j < N; j++) {
      orbitals[j] = (unsigned int)(rand() % 2);
    }
    coef = (rand() / (double)RAND_MAX) +
           (rand() / (double)RAND_MAX) * (double complex)I;
    sdet = slater_determinant_init_c(N, coef, orbitals);
    wavefunction_append_slater_determinant_c(wfn, sdet);
  }

  norm = wavefunction_norm_c(wfn);
  wfn = wavefunction_scalar_multiplication_c(wfn, 1 / norm);

  printf("Wavefunction initialized.\n");

  for (i = 0; i < num_operations; i++) {
    pauli_sums[i] = pauli_sum_init_c(N);

    for (j = 0; j < 3; j++) {
      for (k = 0; k < N; k++) {
        pauli_type = (unsigned int)(rand() % 4);
        paulis[k] = pauli_type;
      }
      coef = (rand() / (double)RAND_MAX) * (double complex)I;
      pString = pauli_string_init_as_ints_c(N, coef, paulis);
      pauli_sum_append_pauli_string_c(pauli_sums[i], pString);
    }
  }

  printf("Pauli sums generated.\n");

  old_wfns = (WavefunctionC **)malloc(num_operations * sizeof(WavefunctionC *));

  start_time = clock();

  for (i = 0; i < num_operations; i++) {
    pSum_str = pauli_sum_to_string_c(pauli_sums[i]);
    printf("%u: %s\n", i, pSum_str);
    free(pSum_str);

    new_wfn = wavefunction_pauli_sum_evolution_c(pauli_sums[i], wfn, 0.01);
    old_wfns[i] = wfn;
    wfn = new_wfn;
  }

  end_time = clock();
  elapsed_time = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;

  printf("Completed %u Pauli sum multiplications in %.4f seconds.\n",
         num_operations, elapsed_time);
  printf("Final wavefunction norm: %lf\n", wavefunction_norm_c(wfn));
  printf(
      "Final number of Slater determinants is %u showing a 2^%.2f increase.\n",
      wfn->s, log2((double)wfn->s / num_sdets));

  free_wavefunction_c(wfn);
  for (i = 0; i < num_operations; i++) {
    free_wavefunction_c(old_wfns[i]);
    free_pauli_sum_c(pauli_sums[i]);
  }
  free(old_wfns);
}
int main(void) {
  printf("\nTesting Initialization, Scaling, and Freeing\n");
  test_initialization_scaling_freeing();
  printf("\n\nTesting Inner Products\n");
  test_inner_product();
  printf("\n\nTesting Appending Slater Determinants\n");
  test_appending_slater_determinants();
  printf("\n\nTesting Wavefunction Multiplication by Pauli Sum\n");
  test_wavefunction_pauli_sum_multiplication();
  printf("\n\nTesting Wavefunction Evolution by Pauli String\n");
  test_wavefunction_pauli_string_evolution();
  printf("\n\nTesting Wavefunction Evolution by Pauli Sum\n");
  test_wavefunction_pauli_sum_evolution();
  printf("\n\nTesting Wavefunction Speed");
  test_wavefunction_speed();
}

File: tests/test_wavefunction.py
from math import log2
import random
import time

from sparse_sim.sparse_sim import *


def test_initialization_scaling_freeing():
    orbitals0 = [0, 1, 0, 1]
    sdet0 = SlaterDeterminant(4, 1 + 0j, orbitals0)

    orbitals1 = [1, 0, 1, 0]
    sdet1 = SlaterDeterminant(4, 1j, orbitals1)

    wfn = Wavefunction(4)

    wfn.append_slater_determinant(sdet0)
    wfn.append_slater_determinant(sdet1)

    print(
        f"Wavefunction initialized with {wfn.s} sdets: {sdet0}, {sdet1}\n")

    print(wfn)

    norm = wfn.norm()
    print(f"Norm: {norm}")

    normalized_wfn = (1 / norm) * wfn
    print(f"Normalized: {normalized_wfn}")

    adjoint_wfn = normalized_wfn.adjoint()
    print(f"Adjoint: {adjoint_wfn}")


def test_inner_product():
    orbitals0 = [0, 1, 0, 1]
    sdet0 = SlaterDeterminant(4, 1 + 0j, orbitals0)

    orbitals1 = [1, 0, 1, 0]
    sdet1 = SlaterDeterminant(4, 1j, orbitals1)

    ket = Wavefunction(4)
    ket.append_slater_determinant(sdet0)
    ket.append_slater_determinant(sdet1)

    bra = ket.adjoint()

    inner_prod = bra * ket

    print(f"({bra}) * ({ket}) = {inner_prod}")


def test_appending_slater_determinants():
    print("Order of Adding:")

    wfn1 = Wavefunction(4)

    for i in range(2):
        for j in range(2):
            for k in range(2):
                for l in range(2):
                    orbitals = [i, j, k, l]
                    sdet = SlaterDeterminant(4, 1 + 0j, orbitals)
                    wfn1.append_slater_determinant(sdet)

    wfn2 = Wavefunction(4)

    for i in range(2):
        for j in range(2):
            for k in range(2):
                for l in range(2):
                    orbitals = [l, k, j, i]
                    sdet = SlaterDeterminant(4, 1 + 0j, orbitals)
                    wfn2.append_slater_determinant(sdet)

    print(wfn1)
    print()
    print(wfn2)

    wfn1 = wfn1.adjoint()

    product = wfn1 * wfn2
    print(f"Product: {product}")


def test_wavefunction_pauli_sum_multiplication():
    orbitals0 = [0, 1, 0, 1]
    orbitals1 = [1, 0, 1, 0]

    sdet0 = SlaterDeterminant(4, 1 + 0j, orbitals0)
    sdet1 = SlaterDeterminant(4, 1 + 0j, orbitals1)

    wfn = Wavefunction(4)
    wfn.append_slater_determinant(sdet0)
    wfn.append_slater_determinant(sdet1)

    paulis0 = ["I", "Y", "I", "X"]
    pString0 = PauliString(4, 1 + 0j, paulis0)

    paulis1 = ["X", "I", "X", "I"]
    pString1 = PauliString(4, 1 + 0j, paulis1)

    pSum = PauliSum(4)
    pSum.append_pauli_string(pString0)
    pSum.append_pauli_string(pString1)

    new_wfn = pSum * wfn

    print(f"({pSum}) * ({wfn}) = {new_wfn}")


def test_wavefunction_pauli_string_evolution():
    orbitals0 = [0, 1, 0, 1]
    sdet0 = SlaterDeterminant(4, 1 + 0j, orbitals0)

    wfn = Wavefunction(4)
    wfn.append_slater_determinant(sdet0)

    paulis0 = ["X", "X", "X", "X"]
    pString = PauliString(4, 1j, paulis0)

    old_wfn = wfn

    for _ in range(100000):
        wfn = wavefunction_pauli_string_evolution(pString, wfn, 0.01)

    norm = wfn.norm()
    print(f"exp(0.01 * ({pString})) * ({old_wfn}) = {wfn}")
    print(f"Norm: {norm}")


def test_wavefunction_pauli_sum_evolution():
    orbitals0 = [0, 1, 0, 1]
    sdet0 = SlaterDeterminant(4, 1 + 0j, orbitals0)

    wfn = Wavefunction(4)
    wfn.append_slater_determinant(sdet0)

    paulis0 = ["X", "X", "X", "X"]
    pString0 = PauliString(4, 0.5j, paulis0)
    pString1 = PauliString(4, 0.5j, paulis0)

    pSum = PauliSum(4)
    pSum.append_pauli_string(pString0)
    pSum.append_pauli_string(pString1)

    old_wfn = wfn

    for _ in range(100000):
        wfn = wavefunction_pauli_sum_evolution(pSum, wfn, 0.01)

    norm = wfn.norm()
    print(f"exp(0.01 * ({pSum})) * ({old_wfn}) = {wfn}")
    print(f"Norm: {norm}")


def test_wavefunction_speed():
    N = 20  # Number of qubits
    num_sdets = 2  # Number of Slater determinants
    num_operations = 20  # Number of Pauli sum operations

    print(
        f"\nRunning Wavefunction Speed Test with {N} qubits, {num_sdets} Slater determinants, {num_operations} Pauli sum operations.")

    # Step 1: Initialize a large wavefunction
    wfn = Wavefunction(N)

    for _ in range(num_sdets):
        orbitals = [random.randint(0, 1) for _ in range(N)]
        coef = random.random() + random.random() * 1j
        sdet = SlaterDeterminant(N, coef, orbitals)
        wfn.append_slater_determinant(sdet)

    wfn = wavefunction_scalar_multiplication(wfn, 1 / wfn.norm())

    print("Wavefunction initialized.")

    # Step 2: Generate multiple Pauli sums
    pauli_sums = []
    for _ in range(num_operations):
        pSum = PauliSum(N)
        for _ in range(3):
            paulis = [random.choice(["I", "X", "Y", "Z"]) for _ in range(N)]
            coef = random.random() * 1j
            pString = PauliString(N, coef, paulis)
            pSum.append_pauli_string(pString)
        pauli_sums.append(pSum)

    print("Pauli sums generated.")

    # Step 3: Apply multiple Pauli sum multiplications
    start_time = time.time()

    for i, pSum in enumerate(pauli_sums):
        print(f"{i}: {pSum}")
        wfn = wavefunction_pauli_sum_evolution(pSum, wfn, 0.01)

    end_time = time.time()
    elapsed_time = end_time - start_time

    print(
        f"Completed {num_operations} Pauli sum multiplications in {elapsed_time:.4f} seconds.")
    print(f"Final wavefunction norm: {wfn.norm()}")
    print(
        f"Final number of Slater determinants is {wfn.s} showing a 2^{log2(wfn.s / num_sdets)} increase.")


def test_wavefunction_cleaning():
    orbitals0 = [0, 1, 1, 0]
    sdet0 = SlaterDeterminant(4, 0.5 + 0j, orbitals0)

    orbitals1 = [0, 1, 0, 1]
    sdet1 = SlaterDeterminant(4, 1e-4, orbitals1)

    orbitals2 = [1, 0, 1, 0]
    sdet2 = SlaterDeterminant(4, 1e-4j, orbitals2)

    orbitals0 = [0, 0, 1, 0]
    sdet3 = SlaterDeterminant(4, 0.5 + 0j, orbitals0)

    wfn = sdet0 + sdet1 + sdet2 + sdet3
    print(f"Initial: {wfn}")

    wfn_clean = wfn.remove_near_zero_terms(1e-3)
    print(f"Cleaned: {wfn_clean}")

    prob_dist = wavefunction_to_probability_distribution(wfn_clean)
    print(f"Probability Distribution: {prob_dist}")


def main():
    print("\nTesting Initialization, Scaling, and Freeing")
    test_initialization_scaling_freeing()

    print("\nTesting Inner Products")
    test_inner_product()

    print("\nTesting Appending Slater Determinants")
    test_appending_slater_determinants()

    print("\nTesting Wavefunction Multiplication by Pauli Sum")
    test_wavefunction_pauli_sum_multiplication()

    print("\nTesting Wavefunction Evolution by Pauli String")
    test_wavefunction_pauli_string_evolution()

    print("\nTesting Wavefunction Evolution by Pauli Sum")
    test_wavefunction_pauli_sum_evolution()

    # print("\nTesting Wavefunction Speed")
    # test_wavefunction_speed()

    print("\nTesting Wavefunction Cleaning")
    test_wavefunction_cleaning()


if __name__ == "__main__":
    main()
